### PR TITLE
[REEF-1209] Implement SuspendTask on the Evaluator

### PR DESCRIPTION
**Pull request by afchung**
JIRA:
  [REEF-1209](https://issues.apache.org/jira/browse/REEF-1209)

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Evaluator.Tests/TaskRuntimeTests.cs [hunk 1]
```diff
@@ -214,6 +215,92 @@ public void TestTaskEventsOnFailure()
...
+        public void TestSuspendTask()
+        {
+            var contextId = Guid.NewGuid().ToString();
+            var taskId = Guid.NewGuid().ToString();
+
+            var injector = GetInjector(typeof(CountDownAction), contextId, taskId);
+            var taskRuntime = injector.GetInstance<TaskRuntime>();
+            taskRuntime.RunTask();
+
+            var task = injector.GetInstance<TestTask>();
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Evaluator.Tests/TaskRuntimeTests.cs]: You might want to bind TestTask to ITask and bind TestTask to TaskConfigurationOptions.SuspendHandler, then inject ITask to make the test case clear for what you are using for SuspendHandler explicitly.

File: lang/cs/Org.Apache.REEF.Evaluator.Tests/TaskRuntimeTests.cs [hunk 2]
```diff
@@ -214,6 +215,92 @@ public void TestTaskEventsOnFailure()
...
+        /// Tests that suspend is not invoked after task is done.
+        /// </summary>
+        [Fact]
+        public void TestSuspendTaskAfterDoneIsNotSuspended()
+        {
+            var contextId = Guid.NewGuid().ToString();
+            var taskId = Guid.NewGuid().ToString();
+
+            var injector = GetInjector(contextId, taskId);
+            var taskRuntime = injector.GetInstance<TaskRuntime>();
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Evaluator.Tests/TaskRuntimeTests.cs]: Similar at here, bind TestTask to TaskConfigurationOptions.SuspendHandler explicitly would make the future injection clear for TaskRuntime as an example.

File: lang/cs/Org.Apache.REEF.Evaluator.Tests/TaskRuntimeTests.cs [hunk 3]
```diff
@@ -214,6 +215,92 @@ public void TestTaskEventsOnFailure()
...
+            taskRuntime.RunTask();
+
+            var task = injector.GetInstance<TestTask>();
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Evaluator.Tests/TaskRuntimeTests.cs]: Inject ITask instead of TestTask as you already bound it,

**Comment by jwang98052** [comment]
I completed a pass.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/SuspendEventImpl.cs [hunk 4]
```diff
@@ -20,27 +20,13 @@
...
 {
-    internal sealed class SuspendEventImpl : ICloseEvent
+    internal sealed class SuspendEventImpl : ISuspendEvent
     {
-        public SuspendEventImpl()
-        {
-            Value = Optional<byte[]>.Empty();
-        }
-
         public SuspendEventImpl(byte[] bytes)
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/SuspendEventImpl.cs]: `internal`?

File: lang/cs/Org.Apache.REEF.Common/Tasks/Defaults/DefaultSuspendHandler.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using Org.Apache.REEF.Common.Tasks.Events;
+using Org.Apache.REEF.Tang.Annotations;
+
+namespace Org.Apache.REEF.Common.Tasks.Defaults
+{
+    internal sealed class DefaultSuspendHandler : IObserver<ISuspendEvent>
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Tasks/Defaults/DefaultSuspendHandler.cs]: Please document default behavior in a class-level XML doc.

File: lang/cs/Org.Apache.REEF.Common/Tasks/Events/ISuspendEvent.cs [hunk 6]
```diff
@@ -15,9 +15,12 @@
 // specific language governing permissions and limitations
 // under the License.
 
+using Org.Apache.REEF.Utilities;
+
 namespace Org.Apache.REEF.Common.Tasks.Events
 {
     public interface ISuspendEvent
     {
+        Optional<byte[]> Message { get; }
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Tasks/Events/ISuspendEvent.cs]: Please document. And while you are at it, also document the interface,

File: lang/cs/Org.Apache.REEF.Common/Tasks/Exceptions/TaskSuspendHandlerException.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+
+namespace Org.Apache.REEF.Common.Tasks.Exceptions
+{
+    internal sealed class TaskSuspendHandlerException : Exception
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Tasks/Exceptions/TaskSuspendHandlerException.cs]: Document.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/SuspendEventImpl.cs (hunk 4 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/SuspendEventImpl.cs]: the class is internal.

**Comment by markusweimer** [comment]
looks good to me.

@jwang98052 can we merge?

**Pull request closed without merging**